h2o-persist-s3/src/main/java/water/persist/PersistS3.java

```diff
@@ -35,13 +39,41 @@
 
   private static final String KEY_PREFIX = "s3://";
   private static final int KEY_PREFIX_LEN = KEY_PREFIX.length();
+  private static final int LOCK_TIMEOUT_SEC = 60;
 
-  private static final Object _lock = new Object();
+  private static final Lock _lock = new ReentrantLock();
+  
   private static volatile AmazonS3 _s3;
 
+  public static void changeClientCredentials(final String accesKeyId, final String accessSecretKey) {
```

Comment:
another thought: all the locking seems to be done for Cache - if we separate all the caching to a new object, we can just swap the object to a new instance and avoid any locking

I believe there is a simpler approach that doesn't require locking

Reply:
There is no need to have locking mechanism at all in this method. This was my mistake. Once the cache is updated, it has a local copy of the pointer to the S3 client unless the operation is finished. Any subsequent  calls to this method will NOT trigger the update. Which means if the client is swapped in the middle of an update, it will still finished properly.

Having the caches in separate  class doesn't help anything. Not without proper cache deprecation mechanism, which in fact requires checks on current amazon S3 client or event-based deprecation. Which again requires some degree of locking.

This method creates short-lived caches. Those are linked to a bucket and a bucket is linked to credentials. Changing credentials mostly means accessing a different bucket (typical use case). This renders the old cache obsolete, as it is meant to be short-lived and only prevent DoS attack on Amazon while the user is typing. This means it is perfectly OK to delete the caches once new client is created.

**So I completely removed the lock.**

There is one corner case when the user hits `Import file` button in Flow and changes S3 credentials between the file search phase and hitting the `Import file` button. This can be hardly prevent and the error is very clear:

![snimek z 2019-01-22 11-12-39](https://user-images.githubusercontent.com/8769110/51528426-ef197100-1e36-11e9-97b6-3de0a62094de.png)